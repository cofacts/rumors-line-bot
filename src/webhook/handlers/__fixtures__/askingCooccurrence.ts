import {
  ListArticlesInInitStateQuery,
  ListArticlesInProcessMediaQuery,
} from 'typegen/graphql';

export const textMessage = {
  id: 'text-message',
  text: 'This is a text message',
  articleType: 'TEXT',
  attachmentUrl: null,
  replyCount: 0,
} as const satisfies NonNullable<
  ListArticlesInInitStateQuery['ListArticles']
>['edges'][0]['node'];

export const imageMessage = {
  id: 'img-message',
  articleType: 'TEXT',
  attachmentUrl: 'https://example.com/image.jpg',
  replyCount: 0,
} as const satisfies NonNullable<
  ListArticlesInProcessMediaQuery['ListArticles']
>['edges'][0]['node'];
