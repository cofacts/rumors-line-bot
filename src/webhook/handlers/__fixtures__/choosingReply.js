export const oneReply = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text: '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
      createdAt: '2018-01-09T05:52:12.658Z',
    },
    GetArticle: {
      replyCount: 1,
    },
  },
};

export const oneReply2 = {
  data: {
    GetReply: {
      type: 'RUMOR',
      text: 'The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply ',
      reference: '',
      createdAt: '2018-01-09T05:52:12.658Z',
    },
    GetArticle: {
      replyCount: 1,
    },
  },
};

export const multipleReplies = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text: '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
      createdAt: '2018-01-09T05:52:12.658Z',
    },
    GetArticle: {
      replyCount: 4,
    },
  },
};
