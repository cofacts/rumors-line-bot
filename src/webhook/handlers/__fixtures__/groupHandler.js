const joinGroup = {
  replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
  type: 'join',
  groupId: 'C4a1',
  webhookEvent: {
    mode: 'active',
    timestamp: 1462629479859,
    source: {
      type: 'group',
      groupId: 'C4a1',
    },
  },
};

const textMessage = {
  replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
  type: 'message',
  groupId: 'C4a1',
  webhookEvent: {
    mode: 'active',
    timestamp: 1462629479859,
    message: {
      type: 'text',
    },
    source: {
      type: 'group',
      groupId: 'C4a1',
    },
  },
};

const expiredTextMessage = {
  replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
  type: 'message',
  groupId: 'C4a1',
  webhookEvent: {
    mode: 'active',
    timestamp: 612921600000,
    message: {
      type: 'text',
    },
    source: {
      type: 'group',
      groupId: 'C4a1',
    },
  },
};

export default { joinGroup, textMessage, expiredTextMessage };
