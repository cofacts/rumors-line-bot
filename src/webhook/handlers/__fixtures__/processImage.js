export const oneImageArticle = {
  data: {
    ListArticles: {
      edges: [
        {
          score: 1.87,
          node: {
            articleType: 'IMAGE',
            attachmentHash: 'ffff8000',
            attachmentUrl: 'http://foo/image.jpeg',
            id: 'image-article-1',
          },
        },
      ],
    },
  },
};

export const oneIdenticalImageArticle = {
  data: {
    ListArticles: {
      edges: [
        {
          score: 2,
          node: {
            articleType: 'IMAGE',
            attachmentHash: 'ffff8000',
            attachmentUrl: 'http://foo/image.jpeg',
            id: 'image-article-1',
          },
        },
      ],
    },
  },
};

export const twelveImageArticles = {
  data: {
    ListArticles: {
      edges: Array.from(Array(6)).reduce(
        arr =>
          arr.concat([
            {
              score: 2,
              node: {
                articleType: 'IMAGE',
                attachmentHash: 'ffff8000',
                attachmentUrl: 'http://foo/image.jpeg',
                id: 'image-article-1',
              },
            },
            {
              score: 1.876656,
              node: {
                articleType: 'IMAGE',
                attachmentHash: 'ffff8001',
                attachmentUrl: 'http://foo/image2.jpeg',
                id: 'image-article-2',
              },
            },
          ]),
        []
      ),
    },
  },
};

export const notFound = {
  data: {
    ListArticles: {
      edges: [],
    },
  },
};
