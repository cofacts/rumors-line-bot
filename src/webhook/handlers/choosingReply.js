import { t } from 'ttag';
import gql from 'src/lib/gql';
import { getArticleURL } from 'src/lib/sharedUtils';
import {
  createReferenceWords,
  createTypeWords,
  ellipsis,
  getLIFFURL,
  ManipulationError,
  FLEX_MESSAGE_ALT_TEXT,
} from './utils';
import ga from 'src/lib/ga';

import UserArticleLink from '../../database/models/userArticleLink';

export default async function choosingReply(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  if (event.type !== 'postback') {
    throw new ManipulationError(t`Please choose from provided options.`);
  }

  const selectedReplyId = (data.selectedReplyId = event.input);

  const { data: getReplyData, errors } = await gql`
    query GetReplyRelatedData($id: String!, $articleId: String!) {
      GetReply(id: $id) {
        type
        text
        reference
        createdAt
      }
      GetArticle(id: $articleId) {
        replyCount
      }
    }
  `({ id: selectedReplyId, articleId: data.selectedArticleId });

  if (errors) {
    throw new ManipulationError(
      t`We have problem retrieving message and reply data, please forward the message again`
    );
  }

  await UserArticleLink.updateTimestamps(userId, data.selectedArticleId, {
    lastRepliedAt: new Date(),
  });

  const { GetReply, GetArticle } = getReplyData;

  const articleUrl = getArticleURL(data.selectedArticleId);
  const typeStr = createTypeWords(GetReply.type).toLocaleLowerCase();
  const helpfulTitle = t`Is the reply helpful?`;

  replies = [
    {
      type: 'text',
      text: `üí° ${t`Someone on the internet replies to the message:`}`,
    },
    {
      type: 'text',
      text: ellipsis(GetReply.text, 2000),
    },
    {
      type: 'text',
      text: ellipsis(createReferenceWords(GetReply), 2000),
    },
    {
      type: 'text',
      text:
        `‚¨ÜÔ∏è ${t`Therefore, the author think the message ${typeStr}.`}\n\n` +
        `üíÅ ${t`These messages are provided by some nice volunteers. Please refer to their references and make judgements on your own.`}\n` +
        (GetArticle.replyCount > 1
          ? `üó£Ô∏è ${t`There are different replies for the message. Read them all here before making judgements:`}\n${articleUrl}\n`
          : '') +
        `\n‚ÅâÔ∏è ${t`If you have different thoughts, you may have your say here:`}\n${articleUrl}`,
    },
    {
      type: 'flex',
      altText: helpfulTitle + '\n' + FLEX_MESSAGE_ALT_TEXT,
      contents: {
        type: 'bubble',
        body: {
          type: 'box',
          layout: 'vertical',
          spacing: 'md',
          paddingAll: 'lg',
          contents: [
            {
              type: 'text',
              wrap: true,
              text: helpfulTitle,
            },
          ],
        },
        footer: {
          type: 'box',
          layout: 'horizontal',
          spacing: 'sm',
          contents: [
            {
              type: 'button',
              style: 'primary',
              color: '#ffb600',
              action: {
                type: 'uri',
                label: t`Yes`,
                uri: getLIFFURL('feedback/yes', userId, data.sessionId),
              },
            },
            {
              type: 'button',
              style: 'primary',
              color: '#ffb600',
              action: {
                type: 'uri',
                label: t`No`,
                uri: getLIFFURL('feedback/no', userId, data.sessionId),
              },
            },
          ],
        },
      },
    },
  ];

  const visitor = ga(userId, state, data.selectedArticleText);
  // Track when user select a reply.
  visitor.event({ ec: 'Reply', ea: 'Selected', el: selectedReplyId });
  // Track which reply type reply to user.
  visitor.event({ ec: 'Reply', ea: 'Type', el: GetReply.type, ni: true });
  visitor.send();

  state = 'ASKING_REPLY_FEEDBACK';

  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
