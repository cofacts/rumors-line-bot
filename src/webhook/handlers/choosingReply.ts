import { t } from 'ttag';
import { z } from 'zod';
import gql from 'src/lib/gql';
import {
  ManipulationError,
  createNotificationSettingsBubble,
  createReplyMessages,
  ellipsis,
} from './utils';
import { getArticleURL, createTypeWords } from 'src/lib/sharedUtils';
import ga from 'src/lib/ga';
import UserSettings from 'src/database/models/userSettings';
import { FlexBubble, Message } from '@line/bot-sdk';
import { ChatbotPostbackHandler } from 'src/types/chatbotState';
import {
  GetReplyRelatedDataQuery,
  GetReplyRelatedDataQueryVariables,
  ReplyTypeEnum,
} from 'typegen/graphql';

const inputSchema = z.object({
  a: z.string().describe('Article ID'),
  r: z.string().describe('Reply ID'),
});

/** Postback input type for CHOOSING_REPLY state handler */
export type Input = z.infer<typeof inputSchema>;

/**
 * @param {string} articleId - Article ID of the article-reply to feedback
 * @param {string} replyId - Reply ID of the article-reply to feedback
 * @returns {object} Flex message bubble object that asks the user if reply is helpful
 */
function createAskReplyFeedbackBubble(
  articleId: string,
  replyId: string
): FlexBubble {
  return {
    type: 'bubble',
    body: {
      type: 'box',
      layout: 'vertical',
      paddingAll: 'xl',
      contents: [
        {
          type: 'text',
          wrap: true,
          text: t`Is the reply helpful?`,
        },
      ],
    },
    footer: {
      type: 'box',
      layout: 'vertical',
      spacing: 'sm',
      contents: [
        {
          type: 'button',
          style: 'primary',
          color: '#00B172',
          action: {
            type: 'uri',
            label: '👍 ' + t`Yes`,
            uri: `${process.env.LIFF_URL}?p=feedback&articleId=${articleId}&replyId=${replyId}&vote=UPVOTE`,
          },
        },
        {
          type: 'button',
          style: 'primary',
          color: '#FB5959',
          action: {
            type: 'uri',
            label: '😕 ' + t`No`,
            uri: `${process.env.LIFF_URL}?p=feedback&articleId=${articleId}&replyId=${replyId}&vote=DOWNVOTE`,
          },
        },
      ],
    },
  };
}

/**
 * @param articleId - article ID to share
 * @param fullArticleText - article text
 * @param replyTypeEnumValue - reply's type enum
 * @returns Flex message bubble object that asks user to share
 */
function createShareBubble(
  articleId: string,
  fullArticleText: string,
  replyTypeEnumValue: ReplyTypeEnum
): FlexBubble {
  const articleUrl = getArticleURL(articleId);
  const articleText = ellipsis(fullArticleText, 15);
  const replyType = createTypeWords(replyTypeEnumValue).toLowerCase();
  const sharedText = t`Someone says the message “${articleText}” ${replyType}.\n\nPlease refer to ${articleUrl} for more information, replies and references.`;
  return {
    type: 'bubble',
    body: {
      type: 'box',
      layout: 'vertical',
      spacing: 'md',
      paddingAll: 'lg',
      contents: [
        {
          type: 'text',
          wrap: true,
          text: t`Don't forget to forward the messages above to others and share with them!`,
        },
      ],
    },
    footer: {
      type: 'box',
      layout: 'vertical',
      spacing: 'sm',
      contents: [
        {
          type: 'button',
          style: 'primary',
          color: '#ffb600',
          action: {
            type: 'uri',
            label: t`Share to friends`,
            uri: `https://line.me/R/msg/text/?${encodeURI(sharedText)}`,
          },
        },
        {
          type: 'button',
          style: 'link',
          color: '#ffb600',
          action: {
            type: 'uri',
            label: t`Provide better reply`,
            uri: getArticleURL(articleId),
          },
        },
      ],
    },
  };
}

const choosingReply: ChatbotPostbackHandler = async ({
  context,
  userId,
  postbackData: { input: postbackInput, state },
}) => {
  let input: Input;
  try {
    input = inputSchema.parse(postbackInput);
  } catch (e) {
    console.error('[choosingReply]', e);
    throw new ManipulationError(t`Please choose from provided options.`);
  }

  const { a: selectedArticleId, r: selectedReplyId } = input;

  const { data: getReplyData, errors } = await gql`
    query GetReplyRelatedData($id: String!, $articleId: String!) {
      GetReply(id: $id) {
        type
        text
        reference
        createdAt
      }
      GetArticle(id: $articleId) {
        text
        replyCount
        createdAt
      }
    }
  `<GetReplyRelatedDataQuery, GetReplyRelatedDataQueryVariables>({
    id: selectedReplyId,
    articleId: selectedArticleId,
  });

  /* istanbul ignore if */
  if (
    errors ||
    getReplyData.GetReply === null ||
    getReplyData.GetArticle === null
  ) {
    console.error('[GetReplyRelatedData]', errors);
    throw new ManipulationError(
      t`We have problem retrieving message and reply data, please forward the message again`
    );
  }

  const { GetReply, GetArticle } = getReplyData;
  const { allowNewReplyUpdate } = await UserSettings.findOrInsertByUserId(
    userId
  );

  const replies: Message[] = [
    ...createReplyMessages(GetReply, GetArticle, selectedArticleId),
    {
      type: 'flex',
      altText: t`Is the reply helpful?`,
      contents: {
        type: 'carousel',
        contents: [
          createAskReplyFeedbackBubble(selectedArticleId, selectedReplyId),

          // Ask user to turn on notification if the user did not turn it on
          process.env.NOTIFY_METHOD &&
            !allowNewReplyUpdate &&
            createNotificationSettingsBubble(),

          createShareBubble(
            selectedArticleId,
            getReplyData.GetArticle.text ?? '',
            GetReply.type
          ),
        ].filter(Boolean),
      },
    },
  ];

  const visitor = ga(userId, state, getReplyData.GetArticle.text ?? '');
  // Track when user select a reply.
  visitor.event({ ec: 'Reply', ea: 'Selected', el: selectedReplyId });
  // Track which reply type reply to user.
  visitor.event({ ec: 'Reply', ea: 'Type', el: GetReply.type, ni: true });
  visitor.send();

  return { context, replies };
};

export default choosingReply;
