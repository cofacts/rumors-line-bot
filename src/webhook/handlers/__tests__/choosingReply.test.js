jest.mock('src/lib/gql');
jest.mock('src/lib/ga');

import MockDate from 'mockdate';
import choosingReply from '../choosingReply';
import * as apiResult from '../__fixtures__/choosingReply';
import gql from 'src/lib/gql';
import ga from 'src/lib/ga';

import UserArticleLink from '../../../database/models/userArticleLink';
import Client from '../../../database/mongoClient';

beforeAll(async () => {
  if (await UserArticleLink.collectionExists()) {
    await (await UserArticleLink.client).drop();
  }
});

afterAll(async () => {
  await (await Client.getInstance()).close();
});

beforeEach(() => {
  MockDate.set('2020-01-01');
  ga.clearAllMocks();
  gql.__reset();
});

afterEach(() => {
  MockDate.reset();
});

// Note: all commented to make unit test pass on other PRs.

describe('should select reply by replyId', () => {
  const params = {
    data: {
      searchedText: '貼圖',
      selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
    },
    state: 'CHOOSING_REPLY',
    event: {
      type: 'postback',
      input: 'AWDZeeV0yCdS-nWhuml8',
      timestamp: 1518964687709,
      postback: {
        data: '{"input":"AWDZeeV0yCdS-nWhuml8","state":"CHOOSING_REPLY"}',
      },
    },
    issuedAt: 1518964688672,
    userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
    replies: [],
    isSkipUser: false,
  };

  it('should handle the case with just one reply', async () => {
    gql.__push(apiResult.oneReply);
    expect(await choosingReply(params)).toMatchSnapshot();
    expect(gql.__finished()).toBe(true);
    expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "ea": "Selected",
            "ec": "Reply",
            "el": "AWDZeeV0yCdS-nWhuml8",
          },
        ],
        Array [
          Object {
            "ea": "Type",
            "ec": "Reply",
            "el": "NOT_ARTICLE",
            "ni": true,
          },
        ],
      ]
    `);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('should handle the case with multiple replies', async () => {
    gql.__push(apiResult.multipleReplies);
    expect(await choosingReply(params)).toMatchSnapshot();
    expect(gql.__finished()).toBe(true);
  });

  it('should update UserArticleLink', async () => {
    gql.__push(apiResult.oneReply);
    await choosingReply(params);
    expect(
      (await UserArticleLink.findByUserId(params.userId)).map(e => ({
        ...e,
        _id: '_id',
      }))
    ).toMatchSnapshot();
  });
});

it('should block non-postback interactions', async () => {
  const params = {
    data: {
      searchedText: '貼圖',
      selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
    },
    state: 'CHOOSING_REPLY',
    event: {
      type: 'text',
      input: '123',
      timestamp: 1518964687709,
    },
    issuedAt: 1518964688672,
    userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
    replies: [],
    isSkipUser: false,
  };

  await expect(choosingReply(params)).rejects.toMatchInlineSnapshot(
    `[Error: Please choose from provided options.]`
  );
});

it('should handle graphql error gracefully', async () => {
  gql.__push({ errors: [] });

  const params = {
    data: {
      searchedText: '貼圖',
      selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
    },
    state: 'CHOOSING_REPLY',
    event: {
      type: 'postback',
      input: 'AWDZeeV0yCdS-nWhuml8',
      timestamp: 1518964687709,
      postback: {
        data: '{"input":"AWDZeeV0yCdS-nWhuml8","state":"CHOOSING_REPLY"}',
      },
    },
    issuedAt: 1518964688672,
    userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
    replies: [],
    isSkipUser: false,
  };

  await expect(choosingReply(params)).rejects.toMatchInlineSnapshot(
    `[Error: We have problem retrieving message and reply data, please forward the message again]`
  );
  expect(gql.__finished()).toBe(true);
});
