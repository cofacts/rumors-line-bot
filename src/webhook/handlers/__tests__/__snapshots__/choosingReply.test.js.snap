// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should select reply by replyId should handle the case with just one reply 1`] = `
Object {
  "data": Object {
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "AWDZeeV0yCdS-nWhuml8",
    "postback": Object {
      "data": "{\\"input\\":\\"AWDZeeV0yCdS-nWhuml8\\",\\"state\\":\\"CHOOSING_REPLY\\"}",
    },
    "timestamp": 1518964687709,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1518964688672,
  "replies": Array [
    Object {
      "text": "💡 Someone on the internet replies to the message:",
      "type": "text",
    },
    Object {
      "text": "這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20",
      "type": "text",
    },
    Object {
      "text": "􀂅 ⚠️️ This reply has no references and it may be biased ⚠️️  􀂅",
      "type": "text",
    },
    Object {
      "text": "⬆️ Therefore, the author think the message invalid request.

💁 These messages are provided by some nice volunteers. Please refer to their references and make judgements on your own.

⁉️ If you have different thoughts, you may have your say here:
https://cofacts.hacktabl.org/article/AWDZYXxAyCdS-nWhumlz",
      "type": "text",
    },
    Object {
      "altText": "Is the reply helpful?
📱 Please proceed on your mobile phone.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Is the reply helpful?",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "paddingAll": "lg",
          "spacing": "md",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "label": "Yes",
                "type": "uri",
                "uri": "https://liff.line.me/1563196602-X6mLdDkW/liff/index.html?p=feedback/yes&token=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJVYWRkYzc0ZGY4YTNhMTc2YjkwMWQ5ZDY0OGIwZmM0ZmUiLCJleHAiOjE1Nzc5MjMyMDB9.rVpljRKoxb65cvss9Rtuup-N9vF0y2n46pQ9SXNRufw",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
            Object {
              "action": Object {
                "label": "No",
                "type": "uri",
                "uri": "https://liff.line.me/1563196602-X6mLdDkW/liff/index.html?p=feedback/no&token=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJVYWRkYzc0ZGY4YTNhMTc2YjkwMWQ5ZDY0OGIwZmM0ZmUiLCJleHAiOjE1Nzc5MjMyMDB9.rVpljRKoxb65cvss9Rtuup-N9vF0y2n46pQ9SXNRufw",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "horizontal",
          "spacing": "sm",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
  ],
  "state": "ASKING_REPLY_FEEDBACK",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`should select reply by replyId should handle the case with multiple replies 1`] = `
Object {
  "data": Object {
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "AWDZeeV0yCdS-nWhuml8",
    "postback": Object {
      "data": "{\\"input\\":\\"AWDZeeV0yCdS-nWhuml8\\",\\"state\\":\\"CHOOSING_REPLY\\"}",
    },
    "timestamp": 1518964687709,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1518964688672,
  "replies": Array [
    Object {
      "text": "💡 Someone on the internet replies to the message:",
      "type": "text",
    },
    Object {
      "text": "這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20",
      "type": "text",
    },
    Object {
      "text": "􀂅 ⚠️️ This reply has no references and it may be biased ⚠️️  􀂅",
      "type": "text",
    },
    Object {
      "text": "⬆️ Therefore, the author think the message invalid request.

💁 These messages are provided by some nice volunteers. Please refer to their references and make judgements on your own.
🗣️ There are different replies for the message. Read them all here before making judgements:
https://cofacts.hacktabl.org/article/AWDZYXxAyCdS-nWhumlz

⁉️ If you have different thoughts, you may have your say here:
https://cofacts.hacktabl.org/article/AWDZYXxAyCdS-nWhumlz",
      "type": "text",
    },
    Object {
      "altText": "Is the reply helpful?
📱 Please proceed on your mobile phone.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Is the reply helpful?",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "paddingAll": "lg",
          "spacing": "md",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "label": "Yes",
                "type": "uri",
                "uri": "https://liff.line.me/1563196602-X6mLdDkW/liff/index.html?p=feedback/yes&token=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJVYWRkYzc0ZGY4YTNhMTc2YjkwMWQ5ZDY0OGIwZmM0ZmUiLCJleHAiOjE1Nzc5MjMyMDB9.rVpljRKoxb65cvss9Rtuup-N9vF0y2n46pQ9SXNRufw",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
            Object {
              "action": Object {
                "label": "No",
                "type": "uri",
                "uri": "https://liff.line.me/1563196602-X6mLdDkW/liff/index.html?p=feedback/no&token=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJVYWRkYzc0ZGY4YTNhMTc2YjkwMWQ5ZDY0OGIwZmM0ZmUiLCJleHAiOjE1Nzc5MjMyMDB9.rVpljRKoxb65cvss9Rtuup-N9vF0y2n46pQ9SXNRufw",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "horizontal",
          "spacing": "sm",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
  ],
  "state": "ASKING_REPLY_FEEDBACK",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`should select reply by replyId should update UserArticleLink 1`] = `
Array [
  Object {
    "_id": "_id",
    "articleId": "AWDZYXxAyCdS-nWhumlz",
    "createdAt": 2020-01-01T00:00:00.000Z,
    "lastRepliedAt": 2020-01-01T00:00:00.000Z,
    "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
  },
]
`;
